_drafts/2024-05-18-get-started-with-aws-tf-01.md

```diff
@@ -0,0 +1,254 @@
+
+If we want to connect to our instance, we need to make our own rules.
+
+When defining a security group rule, we declare if the rule is an ingress or egress rule (incoming or outgoing). We can also declare which port the rule applies to and the ip address range.
+
+We don't want just anyone to connect to this service so for now we'll only allow our local device's public ip address in the address range. You can find this easily using a site like [whatismyipaddress.com](www.whatismyipaddress.com). 
+
+Open `unicon-api.tf` and add the following resources, providing your public ip for `cidr_ipv4` value:
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_security_group" "unicorn_api_security_group" {
+  name = "unicorn-api-security-group"
+}
+
+resource "aws_vpc_security_group_ingress_rule" "allow_from_local_device" {
+  security_group_id = aws_security_group.unicorn_api_security_group.id
+  description       = "allow ingress on any port from any ip"
+  cidr_ipv4         = "<YOUR_IP_ADDRESS_HERE>/32"
+  ip_protocol       = "tcp"
+  from_port         = 22
+  to_port           = 22
+}
+```
+
+Next, apply the security group to the ec2 instance, adding the following property to the ec2 instance:
+
+```terraform 
+# /infra/unicorn-api.tf
+
+resource "aws_instance" "web_service_vm" {
+
+    # ...
+
+  vpc_security_group_ids = [aws_security_group.unicorn_api_security_group.id]
+}
+```
+
+With these new resources defined, we can re-apply our changes with `terraform apply`. This will rebuild the EC2 instance, so we'll need to get the latest ip address so we can ping it.  
+Once the resources have been updated and the latest IP address copied, we can now ping our EC2 instance with:  
+
+`ping <the-ip-address>`
+
+## Step 4 - SSH into the EC2 Instance
+
+In this step we're going to create an ssh key pair using terraform. The public key will be added to the ec2 instance as one of its known ssh public keys.  
+We'll save the private key as a local file so we can reference it when we try connect to the ec2 instance.
+
+First, add the following providers to our `main.tf`:
+
+```terraform
+# infra/main.tf
+
+required_providers {
+
+    # ...
+
+    tls = {
+          source  = "hashicorp/tls"
+          version = "~> 4.0.4"
+        }
+
+    local = {
+      source  = "hashicorp/local"
+      version = "2.4.1"
+    }
+}
+```
+
+Since we've added new providers, we need to run `terraform init` to download their modules.  
+
+Create a new file named `ssh_key.tf` and add the following:
+
+```terraform
+# infra/ssh_key.tf
+
+# create a tls key pair using rsa algorithm
+resource "tls_private_key" "ssh_key" {
+    algorithm = "RSA"
+    rsa_bits = 4096
+}
+
+# save the private key to a local file named "key.pem"
+resource "local_file" "rsa_key" {
+    content = tls_private_key.ssh_key.private_key_openssh
+    filename = "${path.module}/key.pem"
+}
```

Comment:
Similar to your guidance around AWS API keys above, you should add some guidance here explaining that the private key is a credential/secret and therefore needs to be treated like one. eg don't commit to source control, don't share it with anyone and that it provides full access to your EC2 instance.

on an unrelated note, I'd also recommend setting. the `file_permission` property on this resource to `0600` (read-write access to the current user, zero access to anyone else) as the default is `0777` (full access to the file to all users on the system). Doing this should also avoid a potential issue later on in your guide when using this key with the `ssh` command - the `ssh` command will refuse to use any private key if it has open permissions.